src/system/dev/dev_driver.c

```diff
@@ -41,19 +42,29 @@ typedef struct dev_file_arg {
 int dev_read_r(struct _reent* r, void* const arg, uint8_t* buffer, const size_t len) {
 	dev_file_arg_t* file_arg = (dev_file_arg_t*)arg;
 	uint32_t port = file_arg->port;
-	claim_port(port - 1, E_DEVICE_GENERIC);
-	// check flags
-	int32_t result = vexDeviceGenericSerialReceive(device->device_info, buffer, len);
-	return_port(port - 1, result);
+	int32_t recv = 0;
+	do {
+		recv = serial_read(port, (uint8_t*)(buffer + recv), len - recv);
+	} while (!(file_arg->flags & O_NONBLOCK) && recv < 1);
```

Comment:
is it really expected behavior that opening a device like this in blocking mode (which is the default) can starve the calling task when reading or writing? sure, the workaround is to have users open in nonblocking mode and then poll it manually, but is that really what we want? I'd probably implement like

```c
while (
    (recv = serial_read(port, (uint8_t*)(buffer + recv), len - recv) < 1
    && !(file_arg->flags & O_NONBLOCK)
) task_delay(MINIMUM_DELAY_INTERVAL);
```
or something like that

Reply:
(same goes for writing)

Reply:
I couldn't find any constant to serve the purpose of `MINIMUM_DELAY_INTERVAL`, so it now sleep's for one millisecond (it won't sleep at all if it reads data the first time `serial_read` is called) . It is implemented it longform just to check `recv` for `PROS_ERR`. All the same goes for writing.

Reply:
Yeah that was just a placeholder. I think the smallest delay time we use anywhere is 2ms, so maybe make it that

Reply:
Ok, changed to 2ms